torch/csrc/utils/python_arg_parser.cpp

```diff
@@ -544,6 +645,39 @@ bool FunctionSignature::parse(PyObject* args, PyObject* kwargs, PyObject* dst[],
       dst[i++] = args;
       arg_pos = nargs;
       continue;
+    } else if (check_has_torch_function(obj)) {
+      // Collects arguments with __torch_function__ and
+      // in the order in which they should be tried (i.e., skipping redundant types).
+      // by checking for subclasses
```

Comment:
The Python only implementation in `_overrides.py` was really nicely documented and easy to understand. This version is not! It would be great to reference the Python implementation from here (and bulk up the comments), and also add tests to verify that the C++ implementation matches the Python reference implementation.

Reply:
Thank you for this comment, I think I just now caught a bug in the C++ implementation thanks to carefully going through the logic and adding comments. 

I will also add a test that verifies the semantics work as expected for an existing C++ kernel as well as a made-up python kernel (using a made-up one in case one of the existing python implementations gets rewritten in C++ in the future).